examples/cpp/pyperf/PyPerfNativeStackTrace.cc

```diff
@@ -0,0 +1,152 @@
+uint64_t NativeStackTrace::ip = 0;
+
+NativeStackTrace::NativeStackTrace(uint32_t pid, const unsigned char *raw_stack,
+                                   size_t stack_len, uint64_t ip, uint64_t sp) : errors(0) {
+  NativeStackTrace::stack = raw_stack;
+  NativeStackTrace::stack_len = stack_len;
+  NativeStackTrace::ip = ip;
+  NativeStackTrace::sp = sp;
+
+  unw_accessors_t my_accessors = _UPT_accessors;
+  my_accessors.access_mem = NativeStackTrace::access_mem;
+  my_accessors.access_reg = NativeStackTrace::access_reg;
+
+  unw_addr_space_t as = unw_create_addr_space(&my_accessors, 0);
+  void *upt = _UPT_create(pid);
+  if (!upt) {
+    logInfo(1, "Failed to _UPT_create\n");
+    this->symbols.push_back(std::string("[Error]"));
+    errors++;
+    return;
+  }
+
+  unw_cursor_t cursor;
+  int res = unw_init_remote(&cursor, as, upt);
+  if (res) {
+    logInfo(1, "Failed to unw_init_remote: %s\n", unw_strerror(res));
+    this->symbols.push_back(std::string("[Error]"));
+    errors++;
+    return;
+  }
+
+  do {
+    unw_word_t offset;
+    char sym[256];
+
+    if (unw_get_proc_name(&cursor, sym, sizeof(sym), &offset) == 0) {
+      this->symbols.push_back(std::string(sym));
+    } else {
+      unw_word_t ip;
+      unw_get_reg(&cursor, UNW_REG_IP, &ip);
+      logInfo(2,
+              "0x%lx -- error: unable to obtain symbol name for this frame "
+              "(SP=0x%lx)\n",
+              ip, NativeStackTrace::sp);
+      this->symbols.push_back(std::string("(missing)"));
+      errors++;
+      break;
+    }
+
+    if (memcmp(sym, "_PyEval_EvalFrameDefault",
+                sizeof("_PyEval_EvalFrameDefault")) == 0 ||
+        memcmp(sym, "PyEval_EvalFrameEx", sizeof("PyEval_EvalFrameEx")) == 0)
+        {
+      break;
+    }
+  } while (unw_step(&cursor) > 0);
+
+  _UPT_destroy(upt);
+  return;
+}
+
+int NativeStackTrace::access_reg(unw_addr_space_t as, unw_regnum_t regnum,
+                                 unw_word_t *valp, int write, void *arg) {
+  if (regnum == UNW_REG_SP) {
+    if (write) {
+      logInfo(2, "Libunwind attempts to write to SP\n");
+      return -UNW_EINVAL;
+    }
+
+    *valp = NativeStackTrace::sp;
+    return 0;
+  }
+  else if (regnum == UNW_REG_IP) {
+    if (write) {
+      logInfo(2, "Libunwind attempts to write to IP\n");
+      return -UNW_EINVAL;
+    }
+
+    *valp = NativeStackTrace::ip;
+    return 0;
+  }
+  else {
+    logInfo(2, "Libunwind attempts to %s regnum %d\n", write ? "write" : "read", regnum);
+    return -UNW_EBADREG;
+  }
+}
+
+int NativeStackTrace::access_mem(unw_addr_space_t as, unw_word_t addr,
+                                 unw_word_t *valp, int write, void *arg) {
+  if (write) {
+    logInfo(2, "Libunwind mem write attempt\n");
+    return -UNW_EINVAL;
+  }
+
+  const unw_word_t stack_start = NativeStackTrace::sp & ~(getpagesize() - 1);
+  const unw_word_t stack_end = stack_start + NativeStackTrace::stack_len;
+
+  if (addr >= NativeStackTrace::sp && addr < stack_end) {
```

Comment:
Related to https://github.com/Granulate/bcc/pull/24/files#r659360474: why trim at `sp` if you can allow reads for all addrs from `stack_start`?

Reply:
The memory above SP is considered "not allocated" and contains garbage. Libunwind should never attempt to access it, and if it does it indicates that something is wrong and we wouldn't want to allow it.

Reply:
Is it so? I mentioned redzones in the other comment, which are an example to frames that use the stack w/o creating a frame

Reply:
Hmmm it's rare but now that I'm looking for it I actually do see access attempts to addresses below SP.
Nice.